polkadot/node/core/pvf/src/execute/queue.rs

```diff
@@ -278,15 +284,74 @@ impl Queue {
 		}
 
 		let job = queue.remove(job_index).expect("Job is just checked to be in queue; qed");
+		let exec_kind = job.exec_kind;
 
 		if let Some(worker) = worker {
 			assign(self, worker, job);
 		} else {
 			spawn_extra_worker(self, job);
 		}
-		self.metrics.on_execute_kind(priority);
+		self.metrics.on_execute_kind(exec_kind);
 		self.unscheduled.mark_scheduled(priority);
 	}
+
+	async fn update_active_leaves(&mut self, update: ActiveLeavesUpdate, ancestors: Vec<Hash>) {
+		self.prune_deactivated_leaves(&update);
+		self.insert_active_leaf(update, ancestors);
+		self.prune_old_jobs().await;
+	}
+
+	fn prune_deactivated_leaves(&mut self, update: &ActiveLeavesUpdate) {
+		for hash in &update.deactivated {
+			let _ = self.active_leaves.remove(&hash);
+		}
+
+		gum::debug!(target: LOG_TARGET, size = ?self.active_leaves.len(), "Active leaves pruned");
+	}
+
+	fn insert_active_leaf(&mut self, update: ActiveLeavesUpdate, ancestors: Vec<Hash>) {
+		let Some(leaf) = update.activated else { return };
+		let _ = self.active_leaves.insert(leaf.hash, ancestors);
+	}
+
+	async fn prune_old_jobs(&mut self) {
+		for &priority in &[Priority::Backing, Priority::BackingSystemParas] {
+			let Some(queue) = self.unscheduled.get_mut(priority) else { continue };
+			let to_remove: Vec<usize> = queue
+				.iter()
+				.enumerate()
+				.filter_map(|(index, job)| {
+					let relay_parent = match job.exec_kind {
+						PvfExecKind::Backing(x) | PvfExecKind::BackingSystemParas(x) => x,
+						_ => return None,
+					};
+					let in_active_fork = self.active_leaves.iter().any(|(hash, ancestors)| {
+						*hash == relay_parent || ancestors.contains(&relay_parent)
+					});
+					if in_active_fork {
+						None
+					} else {
+						Some(index)
+					}
```

Comment:
We could directly `filter` on the `queue`, so if in active fork we send `ValidationError::ExecutionDeadline` and return `true` otherwise `false`.

Reply:
I can't send the result inside the closure, that why I didn't use retain

Reply:
why not? the send function is not async and we ignore the result